File: packages/shared/src/api/oneMinuteReading.ts
import { getApiClient } from "./client";

/**
 * "Your Chart in One Minute" — docs/ONE_MINUTE_READING_2026-08-04.md.
 *
 * Route verified against the backend decorator before wiring:
 * `@router.get("/charts/{chart_id}/one-minute")` in app/api/charts.py — path
 * param (not query), GET, optional `asOf` query alias. Two wrappers in this
 * package have silently drifted from their routes in the past, so the check is
 * part of adding one.
 */

export interface OneMinuteText {
  ta: string;
  en: string;
}

/**
 * `basis` is the astrological ground for the beat and the ONLY field on this
 * surface allowed to carry technical vocabulary — the body text never does.
 * Render it behind a disclosure, never inline.
 */
export interface OneMinuteBeat {
  id: string;
  text: OneMinuteText;
  basis: OneMinuteText | null;
}

export interface OneMinuteQuestionOption {
  value: string;
  label: OneMinuteText;
}

/**
 * Raised when a beat has been WITHHELD because it turns on a fact we have never
 * asked for — an adult whose `maritalStatus` is null. Answering PATCHes the
 * birth profile, so it improves every other surface too.
 *
 * `beforeBeat` is the id of the beat this question renders above: it stands in
 * the gap the withheld beat left, and the backend owns the beat order, so the
 * anchor is sent rather than guessed. Render every option — the set covers
 * married/single/divorced/widowed, and a reader who is given fewer has to
 * misdescribe themselves to dismiss the question.
 */
export interface OneMinutePendingQuestion {
  field: string;
  prompt: OneMinuteText;
  options: OneMinuteQuestionOption[];
  beforeBeat: string;
}

/** The span the reading is stable for; it moves at the antardasha boundary, never daily. */
export interface OneMinuteReadingWindow {
  from: string;
  to: string;
}

export interface OneMinuteReadingData {
  chartId: string;
  /** Answer `pendingQuestion` against PATCH /birth-profiles/{birthProfileId}. */
  birthProfileId: string;
  displayName: string;
  asOf: string;
  readingWindow: OneMinuteReadingWindow;
  age: number;
  stage: string;
  ageBand: OneMinuteText;
  focusTopic: string;
  /** "self" | "parent" — a minor's reading is addressed to the parent. */
  addressedTo: string;
  beats: OneMinuteBeat[];
  pendingQuestion: OneMinutePendingQuestion | null;
  wordCount: { ta: number; en: number };
  nextStep: { label: OneMinuteText; href: string };
}

export function getOneMinuteReading(
  chartId: string,
  options?: { asOf?: string },
): Promise<{ success: boolean; data: OneMinuteReadingData }> {
  const query = options?.asOf ? `?asOf=${encodeURIComponent(options.asOf)}` : "";
  return getApiClient().get(`/charts/${chartId}/one-minute${query}`) as Promise<{
    success: boolean;
    data: OneMinuteReadingData;
  }>;
}

File: app/schemas/one_minute_reading.py
"""Response shapes for "Your Chart in One Minute" (docs/ONE_MINUTE_READING_2026-08-04.md).

``beats`` is an ARRAY, not seven named fields, deliberately: a beat can be
suppressed (a minor gets no career beat, an unanswered marital-status question
withholds beat 5) without the frontend having to know why. A named-field shape
would force every consumer to re-implement the suppression rules that
``one_minute_reading_service`` already owns.
"""
from __future__ import annotations

from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class OneMinuteText(BaseModel):
    ta: str
    en: str

    model_config = ConfigDict(populate_by_name=True)


class OneMinuteBeat(BaseModel):
    """One paragraph of the reading.

    ``basis`` is the astrological ground for the beat ("Rohini; Saturn
    mahadasha 2015-2022") and is the ONLY field on this surface permitted to
    carry technical vocabulary — the body text never does. Same disclosure
    pattern as the numerology verdict's AlignmentBasis: the plain reader never
    opens it, and the reader who wants to check us can.
    """

    id: str
    text: OneMinuteText
    basis: OneMinuteText | None = None

    model_config = ConfigDict(populate_by_name=True)


class OneMinuteQuestionOption(BaseModel):
    value: str
    label: OneMinuteText

    model_config = ConfigDict(populate_by_name=True)


class OneMinutePendingQuestion(BaseModel):
    """The one question the reading is allowed to ask.

    Raised only when the answer changes which beat 5 is correct and we do not
    hold it — see the service's ``_focus_topic``. Answering PATCHes the birth
    profile, which improves every other surface too.

    ``before_beat`` names the beat the question renders above, and it exists
    because the question now stands in place of a beat that was withheld: with
    nothing to anchor to, the client fell back to the top of the piece and asked
    a reader their marital status before they had read a word. The beat order is
    the service's to know, so the anchor is sent rather than inferred.

    ``options`` is deliberately not a yes/no. "Not yet" is not the same answer as
    "single", and a divorced or widowed reader given only those two buttons has
    to misdescribe themselves to make the question go away.
    """

    field: str
    prompt: OneMinuteText
    options: list[OneMinuteQuestionOption]
    before_beat: str = Field(alias="beforeBeat")

    model_config = ConfigDict(populate_by_name=True)


class OneMinuteReadingWindow(BaseModel):
    """The span this reading is stable for.

    The reading regenerates at the antardasha boundary, never daily — a
    life-story that reads differently each morning announces itself as
    generated. The UI shows this as "as of <month year>".
    """

    from_date: date = Field(alias="from")
    to_date: date = Field(alias="to")

    model_config = ConfigDict(populate_by_name=True)


class OneMinuteWordCount(BaseModel):
    ta: int
    en: int

    model_config = ConfigDict(populate_by_name=True)


class OneMinuteNextStep(BaseModel):
    label: OneMinuteText
    href: str

    model_config = ConfigDict(populate_by_name=True)


class OneMinuteReadingData(BaseModel):
    chart_id: UUID = Field(alias="chartId")
    # Returned so the client can answer `pending_question` against the real
    # PATCH /birth-profiles/{id} route rather than needing a by-chart alias
    # invented for this one surface.
    birth_profile_id: UUID = Field(alias="birthProfileId")
    display_name: str = Field(alias="displayName")
    as_of: date = Field(alias="asOf")
    reading_window: OneMinuteReadingWindow = Field(alias="readingWindow")
    age: int
    stage: str
    age_band: OneMinuteText = Field(alias="ageBand")
    focus_topic: str = Field(alias="focusTopic")
    addressed_to: str = Field(alias="addressedTo")
    beats: list[OneMinuteBeat]
    pending_question: OneMinutePendingQuestion | None = Field(default=None, alias="pendingQuestion")
    word_count: OneMinuteWordCount = Field(alias="wordCount")
    next_step: OneMinuteNextStep = Field(alias="nextStep")

    model_config = ConfigDict(populate_by_name=True)


class OneMinuteMeta(BaseModel):
    calculation_version: str = Field(alias="calculationVersion")
    generated_at: datetime = Field(alias="generatedAt")

    model_config = ConfigDict(populate_by_name=True)


class OneMinuteReadingResponse(BaseModel):
    success: bool = True
    data: OneMinuteReadingData
    meta: OneMinuteMeta

    model_config = ConfigDict(populate_by_name=True)
